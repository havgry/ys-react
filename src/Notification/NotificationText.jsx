import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

const NotificationText = (props) => {
  const {
    children,
    className,
  } = props

  const wrapperProps = {
    className: classNames(
      'ys-notification__text',
      className,
    ),
  }

  return (
    <p {...wrapperProps}>
      {children}
      {/* <a href="#" class="ys-notification__inline-link"></a> */}
    </p>
  )
}

NotificationText.propTypes = {
  children: PropTypes.string.isRequired,
  className: PropTypes.string,
}

NotificationText.defaultProps = {
  className: '',
}

export default NotificationText
