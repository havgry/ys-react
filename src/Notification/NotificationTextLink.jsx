import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import Link from '../Link'

const NotificationTextLink = ({
  children,
  className,
  ...props
}) => {
  const wrapperProps = {
    unstyled: true,
    className: classNames(
      'ys-notification__inline-link',
      className,
    ),
    ...props,
  }

  return (
    <Link {...wrapperProps}>
      {children}
    </Link>
  )
}

NotificationTextLink.propTypes = {
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
}

NotificationTextLink.defaultProps = {
  className: '',
}

export default NotificationTextLink
