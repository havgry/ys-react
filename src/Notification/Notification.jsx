import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import NotificationText from './NotificationText'
import NotificationTextLink from './NotificationTextLink'

import componentPropType from '../helpers/componentPropType'

const Notification = ({
  variant,
  className,
  children,
  ...props
}) => {
  const Icon = () => {
    const variantIcons = {
      info: () => (
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 2 12">
          <g fill="none" fillRule="evenodd">
            <path stroke="currentColor" strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="M1 5v6" />
            <path fill="currentColor" d="M2 1a1 1 0 11-2 0 1 1 0 012 0" />
          </g>
        </svg>
      ),
      success: () => (
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 12 10">
          <path fill="none" stroke="currentColor" strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="M1.5 5.124L5.01 8.75l5.49-7.5" />
        </svg>
      ),
      warning: () => (
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 2 12">
          <g fill="none" fillRule="evenodd">
            <path stroke="currentColor" strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="M1 1v6" />
            <path fill="currentColor" d="M2 11a1 1 0 11-2 0 1 1 0 012 0" />
          </g>
        </svg>
      ),
      error: () => (
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 2 12">
          <g fill="none" fillRule="evenodd">
            <path stroke="currentColor" strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="M1 1v6" />
            <path fill="currentColor" d="M2 11a1 1 0 11-2 0 1 1 0 012 0" />
          </g>
        </svg>
      ),
    }

    const VariantIcon = variantIcons[variant]

    return (
      <VariantIcon className="ys-icon" />
    )
  }

  const wrapperProps = {
    className: classNames(
      'ys-notification',
      `ys-notification--${variant}`,
      className,
    ),
    ...props,
  }

  return (
    <div {...wrapperProps}>
      <span
        className="ys-notification__icon"
        aria-hidden="true"
      >
        <Icon />
      </span>
      <div className="ys-notification__content">
        {children}
      </div>
    </div>
  )
}

Notification.propTypes = {
  children: componentPropType(NotificationText).isRequired,
  variant: PropTypes.oneOf(['info', 'success', 'warning', 'error']),
  className: PropTypes.string,
}

Notification.defaultProps = {
  variant: 'info',
  className: '',
}

Notification.Text = NotificationText
Notification.TextLink = NotificationTextLink

export default Notification
