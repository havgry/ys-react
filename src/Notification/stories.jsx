import React from 'react'
import { storiesOf } from '@storybook/react'

import { select, text } from '@storybook/addon-knobs'

import Notification from './Notification'

const variants = {
  info: 'info',
  success: 'success',
  warning: 'warning',
  error: 'error',
}

storiesOf('Notification', module)
  .addParameters({ jest: ['Notification'] })
  .add('Default', () => (
    <Notification
      variant={select('variant', variants, variants.info, 'Visuals')}
      className={text('className', '', 'Misc')}
    >
      <Notification.Text>{text('text', 'Notification message', 'Content')}</Notification.Text>
    </Notification>
  ))
