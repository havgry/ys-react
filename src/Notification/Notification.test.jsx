import React from 'react'
import { shallow, mount as mountComponent } from 'enzyme'

import Notification from './Notification'
import dummyIcon from '../../utils/dummyIcon'

const defaultMessage = 'Notification message'
const customClassName = 'sparrow'
const getNotification = props => (
  <Notification
    {...props}
  >
    <Notification.Text>
      {defaultMessage}
    </Notification.Text>
  </Notification>
)

const shallowRender = props => shallow(getNotification(props))
const mount = props => mountComponent(getNotification(props))

describe('<Notification>', () => {
  describe('Props', () => {
    it('Doesn\'t throw propType warning if child is `Notification.Text` element', () => {
      const component = () => shallowRender()
      expect(component).not.toThrow()
    })

    it('Doesn\'t throw propType warning if children are multiple `Notification.Text` elements', () => {
      const component = () => shallow(
        <Notification>
          <Notification.Text>{defaultMessage}</Notification.Text>
          <Notification.Text>{defaultMessage}</Notification.Text>
        </Notification>,
      )
      expect(component).not.toThrow()
    })

    it('Throws propType warning if no `child` is passed', () => {
      const component = () => shallow(<Notification />)
      expect(component).toThrow()
    })

    it('Throws propType warning if text is being passed as `child`', () => {
      const component = () => shallow(
        <Notification>
          {defaultMessage}
        </Notification>,
      )
      expect(component).toThrow()
    })

    it('Throws propType warning if unsupported `variant` prop is passed', () => {
      const component = () => shallowRender({
        variant: 'default',
      })

      expect(component).toThrow()
    })

    it('Has correct defaultProps', () => {
      const component = mount()

      const {
        variant,
        // icon,
        displayIcon,
        className,
      } = component.props()

      expect(variant).toBe('info')
      expect(displayIcon).toBe(true)
      expect(className).toBe('')
    })
  })

  describe('Rendering', () => {
    it('Custom `className` is being applied to the container', () => {
      const component = shallowRender({
        className: customClassName,
      })

      expect(component.find(`.ys-notification.${customClassName}`).exists()).toBe(true)
    })

    it('Supports and adds the correct `variant` class to the container', () => {
      const variants = ['info', 'success', 'warning', 'error']

      variants.forEach((variant) => {
        const component = shallowRender({
          variant,
        })

        expect(component.find(`.ys-notification.ys-notification--${variant}`).exists()).toBe(true)
      })
    })

    describe('Icons', () => {
      it('Doesn\'t render the icon container nor classes if `displayIcon` is `false`', () => {
        const component = shallowRender({
          displayIcon: false,
        })

        expect(component.find('.ys-notification__icon').exists()).toBe(false)
      })

      it('Correctly renders icons in the <button> element', () => {
        const component = shallowRender({
          icon: dummyIcon,
        })

        expect(component.find('.ys-notification__icon .ys-icon').exists()).toBe(true)
      })
    })

    describe('Snapshots', () => {
      it('With custom props', () => {
        const component = shallowRender({
          className: customClassName,
          variant: 'success',
          icon: dummyIcon,
        })

        expect(component).toMatchSnapshot()
      })

      it('With default props', () => {
        const component = shallowRender()

        expect(component).toMatchSnapshot()
      })
    })
  })
})
