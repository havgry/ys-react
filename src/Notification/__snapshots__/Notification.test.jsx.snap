// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notification> Snapshots Error variant 1`] = `
<div
  className="ys-notification ys-notification--error"
>
  <span
    aria-hidden="true"
    className="ys-notification__icon"
  >
    <Icon />
  </span>
  <div
    className="ys-notification__content"
  >
    <NotificationText
      className=""
    >
      Notification message
    </NotificationText>
  </div>
</div>
`;

exports[`<Notification> Snapshots Info variant 1`] = `
<div
  className="ys-notification ys-notification--info"
>
  <span
    aria-hidden="true"
    className="ys-notification__icon"
  >
    <Icon />
  </span>
  <div
    className="ys-notification__content"
  >
    <NotificationText
      className=""
    >
      Notification message
    </NotificationText>
  </div>
</div>
`;

exports[`<Notification> Snapshots Success variant 1`] = `
<div
  className="ys-notification ys-notification--success"
>
  <span
    aria-hidden="true"
    className="ys-notification__icon"
  >
    <Icon />
  </span>
  <div
    className="ys-notification__content"
  >
    <NotificationText
      className=""
    >
      Notification message
    </NotificationText>
  </div>
</div>
`;

exports[`<Notification> Snapshots Warning variant 1`] = `
<div
  className="ys-notification ys-notification--warning"
>
  <span
    aria-hidden="true"
    className="ys-notification__icon"
  >
    <Icon />
  </span>
  <div
    className="ys-notification__content"
  >
    <NotificationText
      className=""
    >
      Notification message
    </NotificationText>
  </div>
</div>
`;

exports[`<Notification> Snapshots With custom className 1`] = `
<div
  className="ys-notification ys-notification--info sparrow"
>
  <span
    aria-hidden="true"
    className="ys-notification__icon"
  >
    <Icon />
  </span>
  <div
    className="ys-notification__content"
  >
    <NotificationText
      className=""
    >
      Notification message
    </NotificationText>
  </div>
</div>
`;

exports[`<Notification> Snapshots With default props 1`] = `
<div
  className="ys-notification ys-notification--info"
>
  <span
    aria-hidden="true"
    className="ys-notification__icon"
  >
    <Icon />
  </span>
  <div
    className="ys-notification__content"
  >
    <NotificationText
      className=""
    >
      Notification message
    </NotificationText>
  </div>
</div>
`;
